src/stub/parser.rs

```diff
@@ -1,30 +1,31 @@
 #![allow(clippy::while_let_on_iterator)]
 
-use regex::Regex;
-
 pub mod types;
 pub use types::{Cmd, JoinTerm, JoinTermType, Stub, VariableCommand};
+mod token_stream;
+
+use regex::Regex;
+use token_stream::TokenStream;
 
-pub fn parse_generator_stub(generator: String) -> Stub {
-    let generator = generator.replace('\n', " \n ");
-    let stream = generator.split(' ');
-    Parser::new(stream).parse()
+pub fn parse_generator_stub(generator: &str) -> Stub {
+    let token_stream = TokenStream::from(generator);
+    Parser::new(token_stream).parse()
 }
 
-struct Parser<StreamType: Iterator> {
```

Comment:
TokenStream is not finished, the hack makes testing a lot more difficult due to `replace` returning a string.

Reply:
I fail to see how this was not possible to achieve with the same Parse / TokenStream indirection but just a replace instead of your `next` logic.

Also isn't the `chars` in the `next` creating a whole iterator each time you call `next`?